packages/synapse-react-client/src/components/DataGrid/hooks/useGridUndo.tsx

```diff
@@ -0,0 +1,206 @@
+import { DataGridRow, GridModel, Operation } from '../DataGridTypes'
+import { useCallback, useEffect, useState } from 'react'
+import { ModelChange } from '../utils/applyModelChange'
+import { Button, Menu, MenuItem } from '@mui/material'
+import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown'
+import { getInverseOperation } from '../utils/getInverseOperation'
+import { useStack } from './useStack'
+
+/**
+ * Represents a single undoable action in the grid
+ */
+export type UndoableAction = {
+  type: 'CREATE' | 'DELETE' | 'UPDATE'
+  rowIndex: number
+  previousValue?: DataGridRow
+  newValue?: DataGridRow
+}
+
+/**
+ * Hook to manage undo functionality for a grid backed by a CRDT model.
+ *
+ * @param model - current GridModel instance
+ * @param onApplyModelChange - Function to apply model changes and commit
+ * @param onUndo - Optional callback invoked after undoing changes
+ */
+export function useGridUndo(
+  model: GridModel | null,
+  onApplyModelChange: (change: ModelChange) => void,
+  onUndo?: () => void,
+) {
+  // Tracks the stack of undoable actions in LIFO order
+  const undoStack = useStack<UndoableAction>([], 100) // 100 = max undo steps
```

Comment:
* Extract magic number as a constant with a self-documenting name
* Remove `onUndo` since I don't think we need it anymore (see comments on `SynapseGrid`)

```suggestion
const MAX_UNDO_STEPS = 100


/**
 * Hook to manage undo functionality for a grid backed by a CRDT model.
 *
 * @param model - current GridModel instance
 * @param onApplyModelChange - Function to apply model changes and commit
 * @param onUndo - Optional callback invoked after undoing changes
 */
export function useGridUndo(
  model: GridModel | null,
  onApplyModelChange: (change: ModelChange) => void
) {
  // Tracks the stack of undoable actions in LIFO order
  const undoStack = useStack<UndoableAction>([], MAX_UNDO_STEPS)
```